airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/abstract_stream.py

```diff
@@ -0,0 +1,151 @@
+    This interface is not yet stable and may change in the future. Use at your own risk.
+
+    Why create a new interface instead of adding concurrency capabilities the existing Stream?
+    We learnt a lot since the initial design of the Stream interface, and we wanted to take the opportunity to improve.
+
+    High level, the changes we are targeting are:
+    - Removing superfluous or leaky parameters from the methods' interfaces
+    - Using composition instead of inheritance to add new capabilities
+
+    To allow us to iterate fast while ensuring backwards compatibility, we are creating a new interface with a facade object that will bridge the old and the new interfaces.
+    Source connectors that which to leverage concurrency need to implement this new interface. An example will be available shortly
+
+    Current restrictions on sources that implement this interface. Not all of these restrictions will be lifted in the future, but most will as we iterate on the design.
+    - Only full refresh is supported. This will be addressed in the future.
+    - The read method does not accept a cursor_field. Streams must be internally aware of the cursor field to use. User-defined cursor fields can be implemented by modifying the connector's main method to instantiate the streams with the configured cursor field.
+    - Streams cannot return user-friendly messages by overriding Stream.get_error_display_message. This will be addressed in the future.
+    - The Stream's behavior cannot depend on a namespace
+    - TypeTransformer is not supported. This will be addressed in the future.
+    """
+
+    @abstractmethod
+    def read(self) -> Iterable[StreamData]:
+        """
+        Read a stream in full refresh mode
+        :return: The stream's records
+        """
+
+    @property
+    @abstractmethod
+    def name(self) -> str:
+        """
+        :return: Stream name. By default, this is the implementing class name, but it can be overridden as needed.
+        """
+
+    @property
+    def logger(self) -> logging.Logger:
+        return logging.getLogger(f"airbyte.streams.{self.name}")
+
+    @property
+    @abstractmethod
+    def cursor_field(self) -> Union[str, List[str]]:
+        """
+        Override to return the default cursor field used by this stream e.g: an API entity might always use created_at as the cursor field.
+        :return: The name of the field used as a cursor. If the cursor is nested, return an array consisting of the path to the cursor.
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def primary_key(self) -> Optional[Union[str, List[str], List[List[str]]]]:
+        """
+        :return: string if single primary key, list of strings if composite primary key, list of list of strings if composite primary key consisting of nested fields.
+          If the stream has no primary keys, return None.
+        """
```

Comment:
nitpick, could use a typedef like
```python
FieldPath = Union[str, List[str]]

primary_key(self) -> Optional[Union[FieldPath, List[FieldPath]]
```

Because the signature misses the case when we have a composite key with one root field and one nested. WDYT?

Reply:
Agreed on using a typedef, but I defined primary_key as `FieldPath` since we're deprecating support for nested primary key for most (all?) destinations. They are still supported by the protocol, but this will help enforce best practices